File: pset6/credit/credit.c
#include <stdio.h>
#include <cs50.h>
#include <math.h>

long get_cc_number(string prompt);
string check_cc_num_validity(long);
int even_digis(long, int);
int odd_digis(long, int);
string check_validity_math(long, int, string);

int main(void)
{
// Get credit card number from user
    long cc_num = get_cc_number("Please enter a credit card number\n");

// Test cc number for validity
    string valid = check_cc_num_validity(cc_num);
    printf("%s\n", valid);
}


// Prompt user for credit card number
long get_cc_number(string prompt)
{
    long cc_num;
    do
    {
        cc_num = get_long("%s", prompt);
    }
    while (cc_num < 1);
    return cc_num;
}

// Test credit card number for validity based on
// first two numbers and length of credit card number
string check_cc_num_validity(long cc_num)
{
    // Get length of the cc # and the first & first two digits
    int len = log10(cc_num) + 1;
    int first_digi = cc_num / pow(10, len - 1);
    int first_two = cc_num / pow(10, len - 2);

    // initialize valid variable
    string valid = "";

    // first check validity of cc # based on the
    // first 1-2 digits & cc # length.
    // then check math based on Luhn's Algorithm
    if ((first_two == 34) || (first_two == 37))
    {
        if (len != 15)
        {
            valid = "INVALID";
        }
        else
        {
            string cc_type = "AMEX";
            valid = check_validity_math(cc_num, len, cc_type);
        }
    }
    else if (first_digi == 4)
    {
        if ((len != 13) && (len != 16))
        {
            valid = "INVALID";
        }
        else
        {
            string cc_type = "VISA";
            valid = check_validity_math(cc_num, len, cc_type);
        }
    }
    else if ((first_two >= 51) && (first_two <= 55))
    {
        if (len != 16)
        {
            valid = "INVALID";
        }
        else
        {
            string cc_type = "MASTERCARD";
            valid = check_validity_math(cc_num, len, cc_type);
        }
    }
    else
    {
        valid = "INVALID";
    }
    return valid;
}

string check_validity_math(long cc_num, int len, string cc_type)
{
    int even_digi_sum = even_digis(cc_num, len);
    int odd_digi_sum = odd_digis(cc_num, len);

    int total_sum = even_digi_sum + odd_digi_sum;
    int last_digi = total_sum % 10;
    string validity = "";

    if (last_digi == 0)
    {
        validity = cc_type;
    }
    else
    {
        validity = "INVALID";
    }
    return validity;
}

int even_digis(long cc_num, int len)
{
    // initialize the digit sum
    int even_digi_sum = 0;

    // set up loop to run through every other digit,
    // starting with the penultimate
    for (int i = 1; i < len; i = i + 2)
    {
        // get the desired digit to the "ones" position
        // by dividing by the power of 10 for the desired digit
        long digi = (cc_num / pow(10, i));

        // get the desired digit by returning the mod
        // of the transformed number divided by 10
        digi = digi % 10;

        // special case for digits over 5 since their value * 2
        // will be two digits long
        if (digi < 5)
        {
            digi = digi * 2;
            even_digi_sum = even_digi_sum + digi;
        }
        else
        {
            digi = (digi - 5) * 2 + 1;
            even_digi_sum = even_digi_sum + digi;
        }
    }
    return even_digi_sum;
}

int odd_digis(long cc_num, int len)
{
    // initialize the digit sum
    int odd_digi_sum = 0;

    // set up loop to run through every other digit,
    // starting with the last
    for (int i = 0; i < len; i = i + 2)
    {
        // get the desired digit to the "ones" position
        // by dividing by the power of 10 for the desired digit
        long digi = (cc_num / pow(10, i));

        // get the desired digit by returning the mod
        // of the transformed number divided by 10
        digi = digi % 10;

        odd_digi_sum = odd_digi_sum + digi;
    }
    return odd_digi_sum;
}

File: pset6/credit/credit.py
from cs50 import get_int, get_string


def main():
    # get a cc number using get_int to make sure the input is in an integer format
    # then convert it to a string for easier manipulation
    cc_num = get_int("Number: ")
    cc_num = str(cc_num)

    # check validity of cc number
    valid = check_cc_num_validity(cc_num)

    # print validity of credit card
    print(f"{valid}")


# function to check the that the initial number check passes
def check_cc_num_validity(cc_num):
    # gets length, first digit, and first two digits of the cc number
    length = len(cc_num)
    first_digi = cc_num[0]
    first_two = cc_num[0:2]

    # checks if the card is an AMEX
    if first_two == '34' or first_two == '37':
        if not length == 15:
            return "INVALID"
        else:
            cc_type = "AMEX"
            valid = check_validity_math(cc_num, length, cc_type)
    # checks if the card is a Visa
    elif first_digi == '4':
        if not length == 13 and not length == 16:
            return "INVALID"
        else:
            cc_type = "VISA"
            valid = check_validity_math(cc_num, length, cc_type)
    # checks if the card is a Mastercard
    elif int(first_two) >= 51 and int(first_two) <= 55:
        if not length == 16:
            return "INVALID"
        else:
            cc_type = "MASTERCARD"
            valid = check_validity_math(cc_num, length, cc_type)
    # returns INVALID after initial digit checks
    else:
        return "INVALID"
    return valid


# function to check the credit card number according to Luhn's Algorithm
def check_validity_math(cc_num, length, cc_type):
    # gets the sum of the even digits in the cc number multiplied by 2
    even_digi_sum = even_digis(cc_num, length)
    # gets the sum of the odd digits in the cc number
    odd_digi_sum = odd_digis(cc_num, length)

    # totals the even digit sum and the odd digit sum
    # then checks if the last digit of the total sum is zero
    # in which case the cc number could plausibly be valid
    total_sum = even_digi_sum + odd_digi_sum
    last_digi = str(total_sum)[-1]

    if last_digi == '0':
        return cc_type
    else:
        return "INVALID"


# function to sum up the even digits
def even_digis(cc_num, length):
    # initialize sum & iterator
    even_digi_sum = 0
    i = -2
    # while loop to run through the cc number
    # this loop runs through the number backwards as the digits start counting from the right
    while i >= -length:
        digi = int(cc_num[i])
        if digi < 5:
            digi = digi * 2
            even_digi_sum = even_digi_sum + digi
        else:
            digi = (digi - 5) * 2 + 1
            even_digi_sum = even_digi_sum + digi
        i = i - 2
    return even_digi_sum


# function to sum up the odd digits
def odd_digis(cc_num, length):
    # initialize sum and iterator
    odd_digi_sum = 0
    i = -1
    # while loop to run through the cc number
    # this loop runs through the number backwards as the digits start counting from the right
    while i >= -length:
        digi = int(cc_num[i])
        odd_digi_sum = odd_digi_sum + digi
        i = i - 2
    return odd_digi_sum


if __name__ == "__main__":
    main()